File: Writing-a-file_solution/mpi-io_solution.f90
PROGRAM mpi_io_prog
  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: wp=KIND(0.0D0)
  INTEGER, PARAMETER :: array_size=1000000
  INTEGER :: ierror
  CHARACTER(len=80) :: filename
  REAL(wp), DIMENSION(array_size) :: data1
  INTEGER, DIMENSION(MPI_STATUS_SIZE) :: status
  INTEGER :: rank, size
  INTEGER :: double_size
  INTEGER :: filehandle
  INTEGER(kind=MPI_OFFSET_KIND) :: offset
  REAL(wp) :: start, end

  CALL mpi_init(ierror)
  CALL mpi_comm_rank(MPI_COMM_WORLD,rank,ierror)
  CALL mpi_comm_size(MPI_COMM_WORLD,size,ierror)

! Initialise data1 array to different values on each process.
  data1 = REAL(rank,wp)

  WRITE(filename,'("test_out_file_mpi_io")')

  CALL mpi_file_open(MPI_COMM_WORLD,filename,MPI_MODE_WRONLY+MPI_MODE_CREATE,&
       &MPI_INFO_NULL,filehandle,ierror)
  CALL mpi_type_size(MPI_DOUBLE_PRECISION,double_size,ierror)
  offset=rank*array_size*double_size
  CALL mpi_file_set_view(filehandle, offset , MPI_DOUBLE_PRECISION,&
       &MPI_DOUBLE_PRECISION, 'native', MPI_INFO_NULL, ierror)

  CALL mpi_barrier(MPI_COMM_WORLD,ierror)
  start = mpi_wtime()
  CALL mpi_file_write_all(filehandle, data1, array_size, MPI_DOUBLE_PRECISION,&
       &status, ierror)
  CALL mpi_barrier(MPI_COMM_WORLD,ierror)
  end = mpi_wtime()

  CALL mpi_file_close(filehandle,ierror)
  IF(rank==0)THEN
     WRITE(6,'("Time taken was ",F10.6)') end-start
  END IF
  CALL mpi_finalize(ierror)

END PROGRAM mpi_io_prog

File: Writing-a-file_solution/mpi-io_solution.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char *argv[])
{
  const int array_size=1000000;
  const int data_tag=21;
  int rank, size, i;
  char filename[80];
  double data1[array_size];
  double start, end;
  MPI_Status status;
  MPI_Offset offset;
  MPI_File fh;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

/* Initialise data1 array to different values on each process. */
  for (i=0; i<array_size; i++)
     data1[i] = (double)rank;

  strcpy(filename,"test_out_file_mpi_io");

  MPI_File_open(MPI_COMM_WORLD, filename, MPI_MODE_WRONLY|MPI_MODE_CREATE,
		MPI_INFO_NULL, &fh);
  offset = rank * array_size * sizeof(double);
  MPI_File_set_view(fh, offset, MPI_DOUBLE, MPI_DOUBLE, "native", MPI_INFO_NULL);

  MPI_Barrier(MPI_COMM_WORLD);
  start = MPI_Wtime();
  MPI_File_write_all(fh, data1, array_size, MPI_DOUBLE, &status);
  MPI_Barrier(MPI_COMM_WORLD);
  end = MPI_Wtime();

  MPI_File_close(&fh);
  if(rank==0){
     printf("Time taken was %g seconds.\n", end-start);
  }
  MPI_Finalize();
  return 0;
}
